src/morx.rs

```diff
@@ -0,0 +1,1824 @@
+
+                            //Subsitute glyphs[start_pos..(end_pos+1)] with ligature
+                            self.glyphs.drain(start_pos..(end_pos + 1));
+
+                            self.glyphs.insert(start_pos, ligature.clone());
+                            i -= end_pos - start_pos; //make adjustment to i after substitution
+
+                            //Push ligature onto stack, only when the next state is non-zero
+                            if self.next_state != 0 {
+                                self.component_stack.push(ligature.clone());
+                            }
+
+                            //"ligature" has been inserted at start_pos in glyphs array.
+                            //And the next glyph in glyphs array will be processed.
+                        }
+
+                        if action & LAST != 0 {
+                            //this is the last action, so exit the loop 'stack
+                            break 'stack;
+                        }
+                    }
+                    //end of loop 'stack
+                }
+                //end of PERFORM_ACTION
+
+                if entry_flags & DONT_ADVANCE == 0 {
+                    break 'glyph; //exit the loop 'glyph unless entry_flags says DONT_ADVANCE
+                } else {
+                    self.component_stack.clear(); //if the entry_flags does say DONT_ADVANCE, then keep looping with the same glyph.
+                                                  //clear the stack
+                }
+            }
+            //end of loop 'glyph
+
+            i += 1; //advance to the next glyph
+        }
+        //end of loop 'glyphs
+
+        Ok(())
+    }
+}
+
+//------------------------------------ NonContextual Lookup ---------------------------------------------------------------
+//This function looks up and returns the noncontexutal substitute of glyph.
+//It returns 0xFFFF for a glyph index out of bounds of the lookup value array indices.
+pub fn noncontextual_lookup<'a>(glyph: u16, lookup_table: &ClassLookupTable<'a>) -> u16 {
+    match lookup(glyph, lookup_table) {
+        None => {
+            return 0xFFFF;
+        }
+        Some(val) => {
+            return val;
+        }
+    }
+}
+
+//------------------------------------ NonContextual Substiution ----------------------------------------------------------
+pub fn noncontextual_substitution<'a>(
+    glyphs: &mut Vec<RawGlyph<()>>,
+    noncontextual_subtable: &NonContextualSubtable<'a>,
+) -> Result<(), ParseError> {
+    let mut glyph: u16;
+    let mut subst: u16;
+    for i in 0..glyphs.len() {
+        glyph = glyphs[i].glyph_index;
+
+        subst = noncontextual_lookup(glyph, &noncontextual_subtable.lookup_table);
+
+        if (subst != 0xFFFF) && (subst != glyph) {
+            glyphs[i].glyph_index = subst;
+            glyphs[i].glyph_origin = GlyphOrigin::Direct;
+        }
+    }
+    Ok(())
```

Comment:
This can be made more idiomatic with an iterator:

```suggestion
    for glyph in glyphs {
        let subst = noncontextual_lookup(glyph.glyph_index, &noncontextual_subtable.lookup_table);

        if (subst != 0xFFFF) && (subst != glyph.glyph_index) {
            glyph.glyph_index = subst;
            glyph.glyph_origin = GlyphOrigin::Direct;
        }
    }
    Ok(())
```